src/opower/utilities/mnzl.py

```diff
@@ -0,0 +1,280 @@
+"""Mercury NZ Limited utility implementation.
+
+This module handles the authentication and API interactions for Mercury NZ.
+It uses OAuth 2.0 with PKCE for secure authentication.
+"""
+
+import base64
+import hashlib
+import json
+import logging
+import secrets
+import ssl
+from typing import Any, Optional, TypedDict
+from urllib.parse import parse_qs, urlparse
+
+import aiohttp
+
+from ..exceptions import InvalidAuth
```

Comment:
It seems you always raise InvalidAuth. This will result to the integration entering the needs reauth state and asking the user to reenter their password. You should only raise it if the password is invalid. For other cases you should raise CannotConnect so that the client can keep retrying.